sklearn/decomposition/_pca.py

```diff
@@ -464,7 +464,7 @@ def _fit_full(self, X, n_components):
             # number of components for which the cumulated explained
             # variance percentage is superior to the desired threshold
             ratio_cumsum = stable_cumsum(explained_variance_ratio_)
-            n_components = np.searchsorted(ratio_cumsum, n_components) + 1
+            n_components = np.nonzero(ratio_cumsum > n_components)[0][0]
```

Comment:
perhaps `np.argmax(ratio_cumsum > n_components)`?

but maybe there's a neater solution still

Reply:
Hi @jnothman! Yes, your solution should work too. I have checked on my end it works perfectly. Also, your solution is much neater! Thank you so much for suggesting that.

Reply:
And I also think it's correct as many tests would be failing, because now lesser n_components are calculated, which messes up output matrix shape, hence many tests are bound to fail